SPECIFICATION.md

```diff
@@ -0,0 +1,154 @@
+non-standard keys, but must prefix them with `X-`, for example `X-Foo: Bar`.
+
+Whitespace around key and value should be stripped by the parsers.
+
+
+### Body
+
+The first empty line separates the header from the body. The body consists of
+glyph definition blocks separated by at least one empty line.
+
+Every glyph definition starts with the glyph declaration, a unicode codepoint
+in square brackets (e.g. `[1F4A9]` for the pile of poo character). It may
+optionally be followed by a UTF-8 representation of the character for easier
+previewing (e.g. `💩`). However, that preview will be ignored by the parser.
+
+The notation used inside the brackets follows the convention described in [*The
+Unicode Standard, Version 10.0, Appendix A, Notational Conventions, Code
+Points*][unicode-10-spec]:
+
+> In running text, an individual Unicode code point is expressed as U+n, where
+> n is four to six hexadecimal digits, using the digits 0–9 and uppercase
+> letters A–F (for 10 through 15, respectively).
+>
+> Leading zeros are omitted, unless the code point would have fewer than four
+> hexadecimal digits—for example, U+0001, U+0012, U+0123, U+1234, U+12345,
+> U+102345.
+>
+> The U+ may be omitted for brevity in tables or when denoting ranges.
+
+The glyph declaration is followed by a list of polylines or glyph references,
+each on a single line.
+
+A polyline looks like this:
+
+    0,0;3,9;6,0
+
+It consists of coordinate pairs, separated by semicolons (`;`). The X and Y
+coordinates are separated with a comma (`,`).  The coordinates are arbitrary
+precision floating point numbers using the full stop (U+002E) as decimal mark.
+If a coordinate is a whole number, the decimal mark and the fractional part of
+the number may be omitted.
+
+The coordinate system starts at the lower left corner. Positive X coordinates
+point to the right. Positive Y coordinates point to the top. Negative
+coordinates are not allowed.
+
+          Y
+    (0,Y) +
+          |
+          |
+          +-----+ X
+      (0,0)   (X,0)
+
+The example above would draw the pointy "outer frame" of the character "A".
+Here is the complete definition of the glyph for the character "LATIN CAPITAL
+LETTER A (U+0041)":
+
+    [0041] A
+    0.8333,2.5;5.1666,2.5
+    0,0;3,9;6,0
+
+Because many glyphs are very similar, references may be used as "includes".
+They start with the "@" symbol (U+0040), followed by a codepoint. Here is the
+definition of the glyph for the character "LATIN CAPITAL LETTER A WITH
+CIRCUMFLEX (U+00C2)":
+
+    [00c2] Â
+    @0041
+    1.5,11.5;3,13;4.5,11.5
+
+An implementation would simply replace the "@0041" part with the expanded
+polylines from the referenced glyph.
+
+Multiple references can be used in a single glyph definition, but they must all
+be on their own line.
+
+To prevent reference loops and to facilitate single-pass parsers, only backward
+references are allowed. All references glyphs must have been previously defined.
```

Comment:
To be honest, I would strongly prefer to support forward references because it allows to sort characters by their codepoint, without any space for interpretation. And when updating one character in the font file, no change of the order is required. If forward references are not allowed, the order of characters is not fully defined, and may even be changed if one character starts to inherit from another character --> bad for version control of the font files.

Reply:
> without any space for interpretation

Could you give an example for "space of interpretation"? What could be mis-interpreted?

> And when updating one character in the font file, no change of the order is required.

A change of order would very rarely be necessary, because unicode already orders characters in a useful way. For example, all latin characters with diacritics are ordered *after* the corresponding latin character.

> bad for version control of the font files

Is this a problem for C files in your experience? C also discourages implicit declarations (it was even invalid in C99).

Reply:
> Could you give an example for "space of interpretation"? What could be mis-interpreted?

Oh, I think "room for interpretation" would be more correct than "space...", but probably you understood me ;)

So, if no forward references are allowed, the order if glyphs in the font file must be chosen according the inheritance order, which is not 100% canonical (if "A" or "B" comes first is not relevant because they don't inherit from each other, so one can swap them if he likes).

But if the glyphs MUST be ordered by their codepage number, it's like a canonical file format - the creator has no room for choosing some arbitrary order. And I **love** canonical file formats xD

> Is this a problem for C files in your experience?

I don't think one can compare this with C files because in C one can use forward declarations / function prototypes...

But let's see the question from another side: Why should we disallow forward references? I think the parser performance and memory usage shouldn't be an issue for the target audience of this font library. And the implementation of forward references is done in a few minutes (see LFF parser). So what's the reason to disallow them?

Reply:
> So what's the reason to disallow them?

The parser is simpler to write, simpler to understand and simpler to maintain.

@rnestler what do you think?